From ca2e42ef4a8588808cff0c6846f28afeb77d7e76 Mon Sep 17 00:00:00 2001
From: Kevin Hao <kexin.hao@windriver.com>
Date: Sat, 18 Apr 2020 20:18:02 +0800
Subject: [PATCH 508/767] PCI: Fix the Cavium ACS quirk

Both the following two commits touched the Cavium ACS quirk function.
    adff286a3be5 ("PCI: Apply Cavium ACS quirk to ThunderX2 and ThunderX3")
    be120d347fc3 ("PCI: quirks : Apply ACS quirk for all devices")

But when trying to resolve these conflicts in merge commit ed197188f380
("Merge branch 'v5.2/standard/base' into v5.2/standard/cn96xx"), the
wrong codes were picked. This patch restores the correct codes.

Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/pci/quirks.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index 5676326d1f8c..470a6d6a1c2e 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -4243,10 +4243,21 @@ static int pci_quirk_amd_sb_acs(struct pci_dev *dev, u16 acs_flags)
 
 static bool pci_quirk_cavium_acs_match(struct pci_dev *dev)
 {
-	if (!pci_is_pcie(dev) || pci_pcie_type(dev) != PCI_EXP_TYPE_ROOT_PORT)
+	if (!pci_is_pcie(dev))
 		return false;
 
-	return (pci_is_pcie(dev) && ((dev->device & 0xf800) == 0xa000));
+	switch (dev->device) {
+	/*
+	 * Effectively selects all downstream ports for whole ThunderX1
+	 * (which represents 8 SoCs).
+	 */
+	case 0xa000 ... 0xa7ff: /* ThunderX1 */
+	case 0xaf84:  /* ThunderX2 */
+	case 0xb884:  /* ThunderX3 */
+		return true;
+	default:
+		return false;
+	}
 }
 
 static int pci_quirk_cavium_acs(struct pci_dev *dev, u16 acs_flags)
-- 
2.31.1

